src/main/java/com/limechain/storage/block/BlockState.java

```diff
@@ -0,0 +1,798 @@
+
+    @Getter
+    private final byte[] genesisHash;
+    @Getter
+    private byte[] lastFinalized;
+
+    /**
+     * Creates a new BlockState instance from genesis
+     * @param repository the kvrepository used to store the block state
+     * @param header the genesis block header
+     */
+    public BlockState(KVRepository<String, Object> repository, BlockHeader header) {
+        this.blockTree = new BlockTree(header);
+        this.db = repository;
+        this.unfinalizedBlocks = new HashMap<>();
+        this.genesisHash = header.getHash();
+        this.lastFinalized = header.getHash();
+
+        setArrivalTime(header.getHash(), Instant.now());
+        setHeader(header);
+        db.save(helper.headerHashKey(header.getBlockNumber()), header.getHash());
+        setBlockBody(header.getHash(), new BlockBody(new ArrayList<>()));
+
+        //set the latest finalized head to the genesis header
+        setFinalizedHash(genesisHash, BigInteger.ZERO, BigInteger.ZERO);
+    }
+
+    /**
+     * Check if the hash is part of the unfinalized blocks in-memory or persisted in the database.
+     *
+     * @param hash the hash of the block header in byte array representation
+     * @return true if the block header is found, false otherwise
+     */
+    public boolean hasHeader(byte[] hash) {
+        if (unfinalizedBlocks.containsKey(hash)) {
+            return true;
+        }
+
+        return hasHeaderInDatabase(hash);
+    }
+
+    /**
+     * Check if the hash is persisted in the database.
+     *
+     * @param hash the hash of the block header in byte array representation
+     * @return true if the block header is found, false otherwise
+     */
+    public boolean hasHeaderInDatabase(byte[] hash) {
+        Optional<Object> foundHeader = db.find(helper.headerKey(hash));
+        return foundHeader.isPresent();
+    }
+
+    /**
+     * Get the block header for a given hash
+     *
+     * @param hash the hash of the block header in byte array representation
+     * @return the block header
+     */
+    public BlockHeader getHeader(byte[] hash) {
+        Block block = unfinalizedBlocks.get(hash);
+        if (block != null && block.getHeader() != null) {
+            return block.getHeader();
+        }
+
+        return loadHeaderFromDatabase(hash);
+    }
+
+    /**
+     * Get the block hash on our best chain with the given number
+     *
+     * @param blockNum the block number
+     * @return the block hash in byte array representation
+     */
+    public byte[] getHashByNumber(BigInteger blockNum) {
+        try {
+            // Try to get the hash from the block tree
+            return blockTree.getHashByNumber(blockNum.longValue());
+        } catch (RuntimeException e) {
+            // If the error is not LowerThanRootException, rethrow it
+            if (!(e instanceof LowerThanRootException)) {
+                throw new RuntimeException("Failed to get hash from blocktree: " + e.getMessage(), e);
```

Comment:
Does this not return the same message twice? Once from the string with `e.getMessage()` and once from the `Throwable` e